fdbclient/NativeAPI.actor.cpp

```diff
@@ -2192,6 +2217,134 @@ ACTOR Future<Void> watchValue(Future<Version> version, Key key, Optional<Value>
+	loop {
+		try {
+			state Reference<WatchMetadata> metadata = cx->getWatchMetadata(key);
+			if (!metadata.isValid()) {
+				return Void();
+			}
+
+			state Future<Version> watchFutureSS = watchValue(Future<Version>(metadata->version), key, metadata->value, cx, metadata->info, metadata->tags);
+			Version watchVersion = wait(watchFutureSS);
+
+			metadata = cx->getWatchMetadata(key);
+			if (!metadata.isValid()) return Void();
+			
+			if (watchVersion >= metadata->version) { // case 1: version_1 (SS) >= version_2 (map)
+				cx->deleteWatchMetadata(key);
+				if(metadata->watchPromise.canBeSet()) metadata->watchPromise.send(watchVersion);
+			} else {
+				if (metadata->watchPromise.getFutureReferenceCount() == 1) { // case 2: version_1 < version_2 and future_count == 1
+					cx->deleteWatchMetadata(key);
+				}
+			}
+		} catch(Error &e) { 
+			if (e.code() == error_code_operation_cancelled) {
+				throw e;
+			}
+			Reference<WatchMetadata> metadata = cx->getWatchMetadata(key);
+			if (!metadata.isValid()) {
+				return Void();
+			} else if (metadata->watchPromise.getFutureReferenceCount() == 1) {
+				cx->deleteWatchMetadata(key);
+				return Void();
+			}
+		}
+	}
+}
+
+ACTOR Future<Void> sameVersionDifKey(Version ver, Key key, Optional<Value> value, Database cx,
+                              TransactionInfo info, TagSet tags) {
+	state ReadYourWritesTransaction tr(cx);
+
+	try {
+
+		tr.setOption( FDBTransactionOptions::READ_SYSTEM_KEYS );
+		state Optional<Value> valSS = wait(tr.get(key));
+		state Reference<WatchMetadata> metadata = cx->getWatchMetadata(key);
+
+		if (metadata.isValid() && valSS != metadata->value) { // val_3 != val_1
+			cx->deleteWatchMetadata(key);
+			
+			metadata->watchPromise.send(ver);
+			metadata->watchFutureSS.cancel();
+		}
+
+		if (valSS == value) { // val_3 == val_2
+			metadata = Reference<WatchMetadata>(new WatchMetadata(value, ver, info, tags));
+			cx->setWatchMetadata(key, metadata);
+
+			metadata->watchFutureSS = watchStorageServerResp(key, cx);
+		}
+
+		if (valSS != value) return Void(); // if val_3 != val_2
+
+		metadata = cx->getWatchMetadata(key);
+		if (metadata.isValid()) { // if val_3 == val_2
+			wait(success(metadata->watchPromise.getFuture()));
+		}
+
+		return Void();
+	}
+	catch(Error& e) {
+		state Error err = e;
+		wait(tr.onError(e));
+		throw err;
+	}
+}
+
+Future<Void> getWatchFuture(Version ver, Key key, Optional<Value> value, Database cx,
+                              TransactionInfo info, TagSet tags) {
+	Reference<WatchMetadata> metadata = cx->getWatchMetadata(key);
+
+	if (!metadata.isValid()) { // case 1: key not in map		
+		metadata = Reference<WatchMetadata>(new WatchMetadata(value, ver, info, tags));
+		cx->setWatchMetadata(key, metadata);
+
+		metadata->watchFutureSS = watchStorageServerResp(key, cx);
+		return success(metadata->watchPromise.getFuture());
+	} 
+	else if (metadata->value == value) { // case 2: val_1 == val_2
```

Comment:
I like these comments describing the cases, but you should define what `val_1` and `val_2` are, and assume anyone reading this code won't have read the design document describing these changes.